PurchasesCoreSwift/Purchasing/PurchasesOrchestrator.swift

```diff
@@ -0,0 +1,508 @@
+                              appUserID: currentAppUserID,
+                              isRestore: isRestore,
+                              productInfo: nil,
+                              presentedOfferingIdentifier: nil,
+                              observerMode: !self.finishTransactions,
+                              subscriberAttributes: unsyncedAttributes) { maybePurchaserInfo, maybeError in
+                self.handleReceiptPost(withPurchaserInfo: maybePurchaserInfo,
+                                       error: maybeError,
+                                       subscriberAttributes: unsyncedAttributes,
+                                       completion: maybeCompletion)
+            }
+        }
+    }
+
+    func handleReceiptPost(withPurchaserInfo maybePurchaserInfo: PurchaserInfo?,
+                           error maybeError: Error?,
+                           subscriberAttributes: SubscriberAttributeDict,
+                           completion maybeCompletion: ReceivePurchaserInfoBlock?) {
+        operationDispatcher.dispatchOnMainThread {
+            if let purchaserInfo = maybePurchaserInfo {
+                self.purchaserInfoManager.cache(purchaserInfo: purchaserInfo, appUserID: self.appUserID)
+            }
+
+            self.markSyncedIfNeeded(subscriberAttributes: subscriberAttributes,
+                                    appUserID: self.appUserID,
+                                    maybeError: maybeError)
+
+            if let completion = maybeCompletion {
+                self.operationDispatcher.dispatchOnMainThread {
+                    completion(maybePurchaserInfo, maybeError)
+                }
+            }
+        }
+    }
+
+    // Although both SKProduct.productIdentifier and SKPayment.productIdentifier
+    // are supposed to be non-null, we've seen instances where this is not true.
+    // so we cast into optionals in order to check nullability, and try to fall back if possible.
+    func extractProductIdentifier(fromProduct product: SKProduct, orPayment payment: SKPayment) -> String? {
+        if let identifierFromProduct = product.productIdentifier as String?,
+           !identifierFromProduct.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
+            return identifierFromProduct
+        }
+        Logger.appleWarning("product.productIdentifier is nil")
+
+        if let identifierFromPayment = payment.productIdentifier as String?,
+           !identifierFromPayment.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
+            return identifierFromPayment
+        }
+        Logger.appleWarning("payment.productIdentifier is nil")
+
+        return nil
+    }
+
+    func preventPurchasePopupCallFromTriggeringCacheRefresh(appUserID: String) {
+        deviceCache.setCacheTimestampToNowToPreventConcurrentPurchaserInfoUpdates(appUserID: appUserID)
+        deviceCache.setOfferingsCacheTimestampToNow()
+    }
+}
+
+// swiftlint:disable:this file_length
```

Comment:
do we have an issue for building out the tests for this class?

Reply:
https://github.com/RevenueCat/purchases-ios/issues/753